diff -Naur src/http/ngx_http_upstream.c src/http3/ngx_http_upstream.c
--- src/http/ngx_http_upstream.c	2022-05-10 17:23:27.000000000 +0800
+++ src/http/ngx_http_upstream.c	2022-05-10 17:22:54.000000000 +0800
@@ -5870,6 +5870,10 @@
 }
 
 
+extern ngx_int_t ngx_http_upstream_dynamic_resolve_directive(ngx_conf_t *cf,
+    ngx_http_upstream_server_t *us, ngx_uint_t *i);
+
+
 static char *
 ngx_http_upstream_server(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
 {
@@ -5878,6 +5882,7 @@
     time_t                       fail_timeout;
     ngx_str_t                   *value, s;
     ngx_url_t                    u;
+    ngx_str_t                    name;
     ngx_int_t                    weight, max_conns, max_fails;
     ngx_uint_t                   i;
     ngx_http_upstream_server_t  *us;
@@ -5891,6 +5896,7 @@
 
     value = cf->args->elts;
 
+    name = value[1];
     weight = 1;
     max_conns = 0;
     max_fails = 1;
@@ -5898,6 +5904,13 @@
 
     for (i = 2; i < cf->args->nelts; i++) {
 
+        ngx_int_t res = ngx_http_upstream_dynamic_resolve_directive(cf, us, &i);
+        if (res == NGX_ERROR) {
+            goto invalid;
+        } else if (res == NGX_AGAIN) {
+            continue;
+        }
+
         if (ngx_strncmp(value[i].data, "weight=", 7) == 0) {
 
             if (!(uscf->flags & NGX_HTTP_UPSTREAM_WEIGHT)) {
@@ -6000,7 +6013,7 @@
         return NGX_CONF_ERROR;
     }
 
-    us->name = u.url;
+    us->name = name;
     us->addrs = u.addrs;
     us->naddrs = u.naddrs;
     us->weight = weight;