diff -Naur src/http/ngx_http_upstream.c src/http3/ngx_http_upstream.c
--- src/http/ngx_http_upstream.c	2021-08-06 09:29:10.000000000 +0800
+++ src/http/ngx_http_upstream.c	2022-05-09 21:53:06.000000000 +0800
@@ -5870,6 +5870,10 @@
 }
 
 
+extern ngx_int_t ngx_http_upstream_dynamic_resolve_directive(ngx_conf_t *cf,
+    ngx_http_upstream_server_t *us, ngx_uint_t *i);
+
+
 static char *
 ngx_http_upstream_server(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
 {
@@ -5898,6 +5902,13 @@
 
     for (i = 2; i < cf->args->nelts; i++) {
 
+        ngx_int_t res = ngx_http_upstream_dynamic_resolve_directive(cf, us, &i);
+        if (res == NGX_ERROR) {
+            goto invalid;
+        } else if (res == NGX_AGAIN) {
+            continue;
+        }
+
         if (ngx_strncmp(value[i].data, "weight=", 7) == 0) {
 
             if (!(uscf->flags & NGX_HTTP_UPSTREAM_WEIGHT)) {
